nixos/modules/config/users-groups.nix

```diff
@@ -624,31 +631,35 @@ in {
       };
     };
 
-    users.groups = {
-      root.gid = ids.gids.root;
-      wheel.gid = ids.gids.wheel;
-      disk.gid = ids.gids.disk;
-      kmem.gid = ids.gids.kmem;
-      tty.gid = ids.gids.tty;
-      floppy.gid = ids.gids.floppy;
-      uucp.gid = ids.gids.uucp;
-      lp.gid = ids.gids.lp;
-      cdrom.gid = ids.gids.cdrom;
-      tape.gid = ids.gids.tape;
-      audio.gid = ids.gids.audio;
-      video.gid = ids.gids.video;
-      dialout.gid = ids.gids.dialout;
-      nogroup.gid = ids.gids.nogroup;
-      users.gid = ids.gids.users;
-      nixbld.gid = ids.gids.nixbld;
-      utmp.gid = ids.gids.utmp;
-      adm.gid = ids.gids.adm;
-      input.gid = ids.gids.input;
-      kvm.gid = ids.gids.kvm;
-      render.gid = ids.gids.render;
-      sgx.gid = ids.gids.sgx;
-      shadow.gid = ids.gids.shadow;
-    };
+    users.groups = mkMerge [
+      {
+        root.gid = ids.gids.root;
+        wheel.gid = ids.gids.wheel;
+        disk.gid = ids.gids.disk;
+        kmem.gid = ids.gids.kmem;
+        tty.gid = ids.gids.tty;
+        floppy.gid = ids.gids.floppy;
+        uucp.gid = ids.gids.uucp;
+        lp.gid = ids.gids.lp;
+        cdrom.gid = ids.gids.cdrom;
+        tape.gid = ids.gids.tape;
+        audio.gid = ids.gids.audio;
+        video.gid = ids.gids.video;
+        dialout.gid = ids.gids.dialout;
+        nogroup.gid = ids.gids.nogroup;
+        users.gid = ids.gids.users;
+        nixbld.gid = ids.gids.nixbld;
+        utmp.gid = ids.gids.utmp;
+        adm.gid = ids.gids.adm;
+        input.gid = ids.gids.input;
+        kvm.gid = ids.gids.kvm;
+        render.gid = ids.gids.render;
+        sgx.gid = ids.gids.sgx;
+        shadow.gid = ids.gids.shadow;
+      }
+      # Create per-user primary groups
+      (lib.mapAttrs' (_: user: { name = user.group; value = { gid = lib.mkDefault user.uid; }; }) nixosConfig.users.users)
```

Comment:
This line seems like a problem to me. If multiple users have the same group, they're all going to set its gid to their own uid. Previously, the group had to be created separately, and would likely have been dynamically allocated. Existing code will begin to fail with this change, because a group shared by multiple users with no explicit gid is now going to have multiple different definitions for its gid.

Even if there aren't multiple users with the group, if the group already existed with a dynamic gid, now there's going to be activation-time warnings saying something like `warning: not applying GID change of group ‘foo’ (998 -> 1000) in /etc/group` because the `mkDefault` adds a new explicit gid.

I think this could be improved if this line only applied to users with the default same-name group. It doesn't fix these issues 100%, because it's possible users had created a group with the same name as their user and introduced these problems with that group. But that's odd enough that I think it's ok, especially since the problems make themselves obvious right away and they're easy to fix.

Reply:
Good catch! How about we guard that line behind `stateVersion >= "23.11"`? (That should have been done to begin with!) If someone explicitly configured same primary group for different users, that should result in eval error (after bumping stateVersion) and the user will have to resolve the gid conflict with `users.groups.mygroup.gid = 123` once.

Does that sound good?

Reply:
> How about we guard that line behind stateVersion >= "23.11"? (That should have been done to begin with!) 

Done now.

Reply:
@bjornfor I still think it's surprising in any case for a group to be automatically assigned the same gid as the user's uid unless it's that user's same-name group. And it's even more surprising if that causes eval failures when multiple users are assigned to the same group.

Reply:
Ok. Then I see a follow-up choice:

* Always create the primary group, but only have the mkDefault gid assignment when user name == primary group. There shouldn't be any eval errors and NixOS should assign the gid dynamically at runtime for groups that don't match a user name.
* Only create the group if user name == primary group.

While testing option 2 I came across what I consider a (security) bug and implemented a potential fix here: https://github.com/NixOS/nixpkgs/pull/258363.